Source/EventFlow.Tests/UnitTests/Sagas/SagaTestClasses.cs

```diff
@@ -0,0 +1,160 @@
+﻿// The MIT License (MIT)
+//
+// Copyright (c) 2015 Rasmus Mikkelsen
+// https://github.com/rasmus/EventFlow
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of
+// this software and associated documentation files (the "Software"), to deal in
+// the Software without restriction, including without limitation the rights to
+// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+using System.Threading;
+using System.Threading.Tasks;
+using EventFlow.Aggregates;
+using EventFlow.Commands;
+using EventFlow.Core;
+using EventFlow.Sagas;
+using EventFlow.ValueObjects;
+
+namespace EventFlow.Tests.UnitTests.Sagas
+{
+    public class SagaTestClasses
+    {
+        public class TestSagaId : SingleValueObject<string>, ISagaId
+        {
+            public TestSagaId(string value) : base(value)
+            {
+            }
+        }
+
+        public class TestSaga : Saga<TestSaga, TestSagaId, TestSagaLocator>,
+            ISagaIsStartedBy<SagaTestAggregate, SagaTestAggregateId, SagaTestEventA>,
+            ISagaHandles<SagaTestAggregate, SagaTestAggregateId, SagaTestEventB>,
+            ISagaHandles<SagaTestAggregate, SagaTestAggregateId, SagaTestEventC>
+        {
+            public TestSaga(TestSagaId id) : base(id)
+            {
+            }
+
+            public Task ProcessAsync(IDomainEvent<SagaTestAggregate, SagaTestAggregateId, SagaTestEventA> domainEvent, CancellationToken cancellationToken)
+            {
+                Schedule(new SagaTestBCommand(domainEvent.AggregateIdentity));
+                Emit(new SagaEventA());
+                return Task.FromResult(0);
+            }
+
+            public Task ProcessAsync(IDomainEvent<SagaTestAggregate, SagaTestAggregateId, SagaTestEventB> domainEvent, CancellationToken cancellationToken)
+            {
+                Schedule(new SagaTestCCommand(domainEvent.AggregateIdentity));
```

Comment:
@MartinHave This was what I was talking about. The problem:

- A saga reacts to a event
- The saga updates and publishes a command (order is irrelevant)
- Command results in another event being published
- As the command bus is synchronous, the saga isn't committed to the event store yet and cannot react correctly to the newly emitted event

The problem would also be present in a world where scheduled command processing is done, although less likely to occur.

So my question is: Does it make sense to "schedule" a series of commands to be published if the saga/aggregate is successfully committed to the event store? Its basically combining an aggregate and synchronous subscriber into one.